test/extensions/filters/http/lua/wrappers_test.cc

```diff
@@ -269,6 +271,50 @@ TEST_F(LuaStreamInfoWrapperTest, ReturnCurrentProtocol) {
   expectToPrintCurrentProtocol(Http::Protocol::Http2);
 }
 
+// Verify downstream addresses are available from stream info wrapper.
+TEST_F(LuaStreamInfoWrapperTest, ReturnCurrentDownstreamAddresses) {
+  const std::string SCRIPT{R"EOF(
+      function callMe(object)
+        testPrint(string.format("%s", object:downstreamLocalAddress()))
+      end
+    )EOF"};
+
+  InSequence s;
+  setup(SCRIPT);
+
+  NiceMock<Envoy::StreamInfo::MockStreamInfo> stream_info;
+  auto address = Network::Address::InstanceConstSharedPtr{
+      new Network::Address::Ipv4Instance("127.0.0.1", 8000)};
+  EXPECT_CALL(stream_info, downstreamLocalAddress()).WillRepeatedly(ReturnRef(address));
+  Filters::Common::Lua::LuaDeathRef<StreamInfoWrapper> wrapper(
+      StreamInfoWrapper::create(coroutine_->luaState(), stream_info), true);
+  EXPECT_CALL(printer_, testPrint(address->asString()));
+  start("callMe");
+  wrapper.reset();
+}
+
+// Verify downstream direct remote addresses are available from stream info wrapper.
+TEST_F(LuaStreamInfoWrapperTest, ReturnCurrentDownstreamAddresses) {
+  const std::string SCRIPT{R"EOF(
+      function callMe(object)
+        testPrint(string.format("%s", object:downstreamDirectRemoteAddress()))
+      end
+    )EOF"};
+
```

Comment:
Could you merge this to the above test?

```lua
function callMe(object)
  testPrint(object:downstreamLocalAddress()))
  testPrint(object: downstreamDirectRemoteAddress())
end
```

And you can mock different address for each, for example:

```
auto downstream_local = Network::Address::InstanceConstSharedPtr{
       new Network::Address::Ipv4Instance("127.0.0.1", 8000)};
auto downstream_direct_remote = Network::Address::InstanceConstSharedPtr{
       new Network::Address::Ipv4Instance("8.8.8.8", 3000)};

```

Reply:
sure, thanks.

Reply:
I tried to merge into one test case, but ran fail, even though mocked a same IP. They all run ok when separated.

I wonder if they can't be mocked together or something wrong? Please take a look. Thanks !

here is the err log:

```
[       OK ] LuaStreamInfoWrapperTest.SetGetComplexDynamicMetadata (1 ms)
                                                                                                                                      117,2         83%
           Actual: "127.0.0.1:0"
         Expected: to be called once
           Actual: called once - saturated and active
test/extensions/filters/http/lua/wrappers_test.cc:253: tried expectation #11: EXPECT_CALL(printer_, testPrint(fmt::format("'{}'", Http::Utility::getProtocolString(protocol.value()))))...
  Expected arg #0: is equal to "'HTTP/1.1'"
           Actual: "127.0.0.1:0"
         Expected: to be called once
           Actual: called once - saturated and active
test/extensions/filters/http/lua/wrappers_test.cc:253: tried expectation #12: EXPECT_CALL(printer_, testPrint(fmt::format("'{}'", Http::Utility::getProtocolString(protocol.value()))))...
  Expected arg #0: is equal to "'HTTP/2'"
           Actual: "127.0.0.1:0"
         Expected: to be called once
           Actual: called once - saturated and active
test/extensions/filters/http/lua/wrappers_test.cc:295: tried expectation #13: EXPECT_CALL(printer_, testPrint(address->asString()))...
  Expected arg #0: is equal to "127.0.0.1:8000"
           Actual: "127.0.0.1:0"
         Expected: to be called once
           Actual: called once - saturated and active
test/extensions/filters/http/lua/wrappers_test.cc:296: tried expectation #14: EXPECT_CALL(printer_, testPrint(downstream_direct_remote->asString()))...
  Expected arg #0: is equal to "127.0.0.1:8000"
           Actual: "127.0.0.1:0"
         Expected: to be called once
           Actual: never called - unsatisfied and active
Stack trace:
  0x2bbbfb9: testing::internal::GoogleTestFailureReporter::ReportFailure()
  0x12450a8: testing::internal::Expect()
  0x2bbe685: testing::internal::UntypedFunctionMockerBase::UntypedInvokeWith()::$_1::operator()()
  0x2bbe2b7: testing::internal::UntypedFunctionMockerBase::UntypedInvokeWith()
  0x1200c74: testing::internal::FunctionMocker<>::Invoke()
  0x1200c25: Envoy::Extensions::Filters::Common::Lua::Printer::testPrint()
  0x122ce43: Envoy::Extensions::Filters::Common::Lua::LuaWrappersTestBase<>::luaTestPrint()
  0x12c7a5a: lj_BC_FUNCC
```

Reply:
Sorry, I think using `ON_CALL` would help:

```diff
diff --git a/test/extensions/filters/http/lua/wrappers_test.cc b/test/extensions/filters/http/lua/wrappers_test.cc
index 81c38f6de..9cdacee4b 100644
--- a/test/extensions/filters/http/lua/wrappers_test.cc
+++ b/test/extensions/filters/http/lua/wrappers_test.cc
@@ -276,8 +276,8 @@ TEST_F(LuaStreamInfoWrapperTest, ReturnCurrentProtocol) {
 TEST_F(LuaStreamInfoWrapperTest, ReturnCurrentDownstreamAddresses) {
   const std::string SCRIPT{R"EOF(
       function callMe(object)
-        testPrint(string.format("%s", object:downstreamLocalAddress()))
-        testPrint(string.format("%s", object:downstreamDirectRemoteAddress()))
+        testPrint(string.format(object:downstreamLocalAddress()))
+        testPrint(string.format(object:downstreamDirectRemoteAddress()))
       end
     )EOF"};
 
@@ -289,9 +289,9 @@ TEST_F(LuaStreamInfoWrapperTest, ReturnCurrentDownstreamAddresses) {
       new Network::Address::Ipv4Instance("127.0.0.1", 8000)};
   auto downstream_direct_remote =
       Network::Address::InstanceConstSharedPtr{new Network::Address::Ipv4Instance("8.8.8.8", 3000)};
-  EXPECT_CALL(stream_info, downstreamLocalAddress()).WillRepeatedly(ReturnRef(address));
-  EXPECT_CALL(stream_info, downstreamDirectRemoteAddress())
-      .WillRepeatedly(ReturnRef(downstream_direct_remote));
+  ON_CALL(stream_info, downstreamLocalAddress()).WillByDefault(ReturnRef(address));
+  ON_CALL(stream_info, downstreamDirectRemoteAddress())
+      .WillByDefault(ReturnRef(downstream_direct_remote));
   Filters::Common::Lua::LuaDeathRef<StreamInfoWrapper> wrapper(
       StreamInfoWrapper::create(coroutine_->luaState(), stream_info), true);
   EXPECT_CALL(printer_, testPrint(address->asString()));
```

Reply:
@dio 
it works!  thanks for your help!  It saves me from going too far on the wrong way.